pkg/events.go

```diff
@@ -438,3 +470,56 @@ func (octopus *EventOctopus) SaveOrUpdate(event Event) error {
 	}
 	return tx.Commit().Error
 }
+
+// recover creates a map from event.Uuid to event.Name
+// for all items in the map that do not have the event.Name == EventCompleted, a new event will be published
+// unless the max retry count has been reached.
+func (octopus *EventOctopus) recover() error {
+	var events []Event
+
+	if err := octopus.Db.Debug().Order("uuid").Find(&events).Error; err != nil {
+		return err
+	}
+
+	// filter out all non-completed events, in-place
+	var added int
+	for i, e := range events {
+		if e.Name != EventCompleted {
+			events[added] = events[i]
+			added++
+		}
+	}
+
+	events = events[0:added]
+
+	publisher, err := octopus.EventPublisher(strings.Replace(octopus.Name, " ", "_", -1))
```

Comment:
whats the replace thing? Can you add a comment?

Reply:
that's a Go inline filtering of a Slice without taking more memory by initiating a new Slice